### ISSUE TITLE
Default values is not respected while parsing null json primitives

### ISSUE THREAD

**Issue by vovkab** [immutables/immutables#341]
Not sure if this is a bug or this is expected behavior.

Let's say I have `Photo` class:

```java
@Value.Immutable
public abstract class Photo {
    @Value.Default()
    @SerializedName("width")
    public int width() {
        return -1;
    }
}
```

Now if we try to parse:

```json
{"width": null}
```

We will crash with:
```
Caused by: com.google.gson.JsonSyntaxException: java.lang.IllegalStateException: Expected NUMBER but was NULL
```

My question is why default value is ignored while parsing json, maybe make sense to skip null and left the rest to the builder?

**Comment by vovkab** [comment]
Looking at default adapter from Gson:
https://github.com/google/gson/blob/master/gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

They do skip all nulls:

```java
while (in.hasNext()) {
    String name = in.nextName();
    BoundField field = boundFields.get(name);
    if (field == null || !field.deserialized) {
        in.skipValue();
    } else {
        field.read(in, instance);
    }
}
```

**Comment by elucash** [comment]
We decided to be strict about that (and our users also demanded it). Think of a `null` is not an absence of value, but rather a value `null` which need to be set. Absence of attribute is when field is not included in JSON. The snippet you provided illustrates something different:

```
while (have more fields) {
   get field name
   find field descriptor by that name
   if (descriptor is not found or found but field is not serializable) {
     skip value
   } else {
     let descriptor read in the next value
   }
}
```

One of the value our Gson binding is a straightforward and strict mapping. I believe the "honest" way of handling it is to define `@Nullable` attribute and another one `@Value.Derived` which is initialized to either supplied value if it's not null or the derived

```java
@Nullable Integer attribute();
@Derived int getAttributeOrDefault() {
   Integer v = attribute();
   return v != null ?  v : 121;
}
```

But I do understand that you probably ending up with some carelessly generated JSON which is not under your control, but you don't want to compromise your model with dirty tricks around null and just want a default value in this case. What I can propose is to add `nullAsDefault = true` attribute to `Gson.TypeAdapters` annotation to generate null-skipping code for default (and non-nullable) attributes.
Another way to do this is to add null-skipping check to generated code and it when `JsonReader.isLenient()` is true (Which can be set using `JsonReader.setLenient(true)` or using `GsonBuilder.setLenient()`.

If you're an advanced Gson hacker you can also subclass `JsonReader` to skip null fields by buffering field name and reading first value token.

**Comment by vovkab** [comment]
@elucash thank you very much for detailed explanation!

`nullAsDefault = true` - sounds good, or maybe even declare adapterType: `STRICT | NULL_SAFE`, this way we can add more if we need different one later.

But I guess my question is - can we achieve desired behavior without making adapter strict?

I think type adapter have to much responsibility and we trying to test backend (by enforcing json structure), also we have multiple entry points to create object. I think we should have just one source of truth - builder (or constructor). And all other data sources should set whatever they have and let builder handle verification.

> Think of a null is not an absence of value, but rather a value null which need to be set.

I like to think about what final result we want to get? Do we need it to fail or we want it to recover?

For example we want builder to fail if we have null returned instead of primitive type. So our strategy would be to skip setting nulls everywhere, this way we can allow builder to take control of the building object instead of adapter.

Do we want to fail if primitive type is null?
 - yes: do nothing, declare it `Integer width()` or `int width()` they both will fail, since we skipped setting null or any value.
 - no: provide default values for both, or make wrapper Nullable.

Maybe I'm missing some edge case we tried to solve with making adapter strict?

**Comment by vovkab** [comment]
@elucash 
Anything I can help with to add `nullAsDefault`?
Do you have contributing guide or some docs about how to start/tests/structure and etc?

**Comment by elucash** [comment]
I'm ok with having `nullAsDefault` but also it worth exploring if we could avoid this knob and sweep it under the rug of `setLenient` flag on `JsonReader` or `GsonBuilder`.

There are no specific contributing guidelines. We are discussing thing in issues and people are making pull requests after that. The coding style is close to a Google style. The main playing parts here is `Gson.TypeAdapters` annotation, it's "mirror" in `GsonMirrors.TypeAdapters`, and the `Gsons.generator` with template syntax. See also `Gsons.java` and `ValueType#gsonTypeAdapters()` as a way to integrate this setting field into template. In theory it should be easy to do this once you familiar with the stuff (and getting familiar, I assume, might not be easy ; )

**Comment by vovkab** [comment]
My first try :D
https://github.com/vovkab/immutables/commit/1efd9b52b16fa53e130afec9fa19db79cf60d23c

**Comment by elucash** [comment]
@vovkab you're implementation is in very right direction. I just ran some micro-benchmarks, on JVM the difference is barely noticeable, but for Android the overhead of checking it for each property should be more visible. Also this should work only for non-nullable default attributes (regarding #345) so decided to go with `nullAsDefault` (it also goes along with other feature attributes as `emptyAsNulls` etc).

Nullable default attributes are introduced in 2.2, but if you don't need them, just not put `@Nullable` on a default attribute. A lot of our users use `@Nullable`, especially on Android, but I and some other folks never use it.

I'm planning to release 2.2.1 tonight or tomorrow (couple of important fixes are underway), but you can get 2.2.1-SNAPSHOT from Sonatype OSS repository updated after each Travis CI build.

```xml
  <repositories>
    <repository>
      <id>sonatype-nexus-snapshots</id>
      <name>Sonatype Nexus Snapshots</name>
      <url>https://oss.sonatype.org/content/repositories/snapshots</url>
      <releases>
        <enabled>false</enabled>
      </releases>
      <snapshots>
        <enabled>true</enabled>
      </snapshots>
    </repository>
  </repositories>
```

**Comment by elucash** [comment]
Released in 2.2.1 Thank you!
